### DISCUSSION TITLE
Quartz scheduled job run twice in the IIS windows server

### DISCUSSION THREAD

**Discussion by parashuramae** [quartznet/quartznet#1489] [Q&A]
Hi All,

We have used Quartz DLL (version : 3.2.4.0) for scheduling the Job in the IIS windows server using the C# code.
The scheduled job trigger twice which is scheduled daily.
How can we resolve the twice run of the job.

Also when the code is modified and released to Production the scheduled job not started as per scheduled time.
The application pool which is used in the Job is recycled. But the job is not triggered. 

Kindly help us to resolve these 2 issue.

Thanks,
Parashurama.

**Comment by lahma** [comment]
You should enable logging and see what is happening.

**Comment by parashuramae** [comment]
Hi Lahma,

Thanks for the response.
We already enabled the logging , but is not giving any info when the scheduled job not started.

Is there any settings / property in the Quartz side we need set to make the scheduled job run only once in the IIS. We have 3 load balance server in the production?

Thanks,
Parashurama.

**Reply by lahma** [reply]
If you have three servers you need to have database and clustered setup to prevent two nodes running same jobs.

**Comment by parashuramae** [comment]
Hi Lahma,

How do we make the clustered setup to prevent two nodes running same jobs?
Do we need to do any settings / property in the Quartz side?

Thanks,
Parashurama.

**Reply by lahma** [reply]
https://github.com/quartznet/quartznet/tree/main/src/Quartz.Examples/13_ClusteringJobsExecution

**Reply by parashuramae** [reply]
Hi Laham,

Thanks for the sharing the  code.
I went through the code and tried the same in our code, but it is giving reference error on the SchedulerBuilder.
I tried to find the DLL for this in the code , but not able to find it in the solution.
Are you using any other way to reference the DLL?

Thanks,
Parashurama.

**Reply by lahma** [reply]
Sorry,  but I don't have the resources to offer step-by-step guidance. I hope you get it sorted.

**Reply by parashuramae** [reply]
Hi Laham,

I was able to get the reference DLL and using the code fixed the twice run issue.
Moved the code with hard coded values for running the Job every 1 hour and released to UAT server/ Production server. Application pool is recycled in all the servers.
The Job is triggered every 1 hour in the UAT server/ Production server and running successfully.

Issue 2:
Also when the code is modified and released to Production the scheduled job not started as per scheduled time.
The application pool which is used in the Job is recycled. But the job is not triggered.

To run the Job on scheduled time we are doing below approach.

1. Store the cron expression in the DB table.
2. Get the cron expression info through the stored procedure(SP).
3. In the code the cron expression is used to trigger the Job on scheduled time.

Once the code is moved to UAT server/Production server the Job is not triggered on scheduled time.

1. To solve this issue the application pool which is used in the Job is recycled in all the servers.
2. But the Job not trigged on scheduled on time.
3. Restarted all the UAT server/ Production server. After the restart of the servers, the Job is not triggered on scheduled on time.

Code : Scheduled time from the DB table

StdSchedulerFactory factory = new StdSchedulerFactory();
IScheduler scheduler = await factory.GetScheduler();
await scheduler.Start();

var ScheduledTime = GetCronInfoForScheduledJob(); //Get the cron expression info through the stored procedure(SP).
if (ScheduledTime != null)
{
IJobDetail DowbloadJob = JobBuilder.Create<DownLoadExcelJob>()
.WithIdentity ("job2" , "group2")
.Build();
IScheduleBuilder DownloadScheduleBuilder = CronScheduleBuilder
.CronSchedule(ScheduledTime)
.InTimeZone(TimeZoneInfo.Utc);

ITrigger trigger1 = TriggerBuilder.Create()
    .WithIdentity("trigger2", "group2")
    .WithSchedule(DownloadScheduleBuilder)
    .ForJob("job2","group2")
    .Build();
 await.= scheduler.ScheduleJob(DowbloadJob, trigger1);
}

We tried with hard coded values in the code for running the Job every 1 hour and below is the code.
Code : For every 1 hour run
ITrigger trigger = TriggerBuilder.Create()
    .WithIdentity("trigger3", "group1")
    .StartNow() 
    .WithSimpleSchedule(x => x
        .WithIntervalInHours(1)
        .RepeatForever())               
    .Build();

Could you please tell us what we are missing in the code / any additional settings in IIS is required to trigger the Job on scheduled time (cron expression) which is taken from the DB table?

Thanks,
Parashurama.

**Reply by lahma** [reply]
Sorry I can't help you when you are using your own custom schedule persistence. Quartz has persistence solution built-in which has been proven to work. This also seems to be very IIS specific.

**Reply by parashuramae** [reply]
Hi Laham,

Thanks for the response.
When you say "IIS specific" means the settings related to IIS in the server.
Are we need to any changes in the IIS settings apart from the app pool recycle?

We also tried with below approach , but is not working.

1. Store the Job schedule time as '1 hour' in the application web.config value.
2. Get the web.config value in the code to trigger the Job on scheduled time.
3. The code is moved to UAT server , the Job is triggered every 1 hour interval & it was successful.
4. After this changed the  web.config value to '4 hour' in the UAT server.
5. But the Job not trigged on scheduled on time.
6. After this the App pool is recycled in all the servers , but the Job is not triggered.

Thanks,
Parashurama.

**Reply by lahma** [reply]
Sorry, I don't know.
